vllm/attention/backends/hpu_attn.py

```diff
@@ -261,6 +284,118 @@ def forward(
         # Reshape the output tensor.
         return output.view(batch_size, seq_len, hidden_size)
 
+    def forward_encoder_decoder(
+        self,
+        query: torch.Tensor,
+        key: torch.Tensor,
+        value: torch.Tensor,
+        kv_cache: torch.Tensor,
+        attn_metadata: HPUAttentionMetadata,
+        k_scale: float = 1.0,
+        v_scale: float = 1.0,
+    ) -> torch.Tensor:
+        """Forward pass with xFormers and PagedAttention.
+
+        Args:
+            query: shape = [num_tokens, num_heads * head_size]
+            key: shape = [num_tokens, num_kv_heads * head_size]
+            value: shape = [num_tokens, num_kv_heads * head_size]
+            kv_cache = [2, num_blocks, block_size * num_kv_heads * head_size]
+            attn_metadata: Metadata for attention.
+        Returns:
+            shape = [num_tokens, num_heads * head_size]
+        """
+        batch_size, _ = query.shape
+
+        if attn_metadata.is_prompt:
+            batch_size = attn_metadata.num_prefills
+            batched_tokens, _ = query.shape
+            batched_kv_tokens, _, _ = key.shape
+            assert batch_size > 0, (
+                "In prefill stage the num_prefills should be > 0")
+            assert batched_tokens % batch_size == 0
+            assert batched_kv_tokens % batch_size == 0
+            seq_len = batched_tokens // batch_size
+
+        hidden_size = self.num_heads * self.head_size
+        query = query.view(-1, self.num_heads, self.head_size)
+        if key is not None:
```

Comment:
is it possible to run forward encoder_decoder with `None` key, value query? 
It is a bit confusing to have multiple checks of `key is not None` or `assert value is None`. 
I guess it is all connected to a logic of running cross-attention vs self-attention.
So in order to make it more readable, could you make it explicit e.g. create a boolean var:
`is_cross_attention = key is None and value is None`
and use this as a condition in the function that uses key and values?

Reply:
This function is only for cross-attention. In the prefill stage, the key & value are not None, but in the decode stage, they are None. There is already a comment in the below code to clarify this (in the line 339).